### ISSUE TITLE
Traverson support in Pact

### ISSUE THREAD

**Issue by okarahan** [DiUS/pact-jvm#100]
Hi,

we use pact to test the contracts between consumer and producer in a microservices architecture.

By now the path to the resources are absolute paths in the contract tests. But in future we want to introduce a traverson to resolve the paths (Giving the resource and resolving the path through traverson). How the traverson is done can be found here:

http://docs.spring.io/spring-hateoas/docs/current/reference/html/#client.traverson

This is also implemented in Spring HATEOAS, which is the library we want to use for traverson:

https://github.com/spring-projects/spring-hateoas

In a few words you can think the traverson of navigating through HAL relations (links). Beginning from the root resource and following the relations, going deeper and deeper until the needed resource is found, e.g.:

```
Traverson traverson = new Traverson(new URI("http://localhost:8080/api/"), MediaTypes.HAL_JSON);
String name = traverson.follow("movies", "movie", "actor").
withTemplateParameters(parameters).
toObject("$.name");
```

Here, the traverser follows HAL relations to find an actor resource and gets the name of the actor.

The problem is, that pact does not support traverson, so it's impossible for us to remove the absolute paths. Do you think it's possible to implement a traverson in pact to resolve the path on 'verifyPact' without the need for an absolute path ?

**Comment by bethesque** [comment]
See: https://github.com/realestate-com-au/pact/wiki/RAQ#doesnt-this-break-hal

We've used HAL a lot in the services that use Pact at realestate.com.au, and I haven't worked out an elegant way to use Pact with hypermedia yet. We've had to use absolute URLs, because when the request gets replayed, it has to know the the URL to replay against.

I have had vague ideas (not yet tried out) of using placeholder URLs in the consumer, or having a way to signify that a particular relation is being mocked rather than a URL, and making the provider create a mapping from the placeholder or relation to an actual URL. I'd be happy to work with you on a proof of concept if you are interested.

**Comment by okarahan** [comment]
Thank you for your response. I talked about this issue with my colleagues but we didn't made a decision yet. I will contact you soon and I think we can make a proof of concept together.

**Comment by uglyog** [comment]
As a work around, we can add some behaviour to the verifyPact task to handle relative URLs, probably by using the base URL for the provider. Would this help?

**Comment by okarahan** [comment]
I don't know at the moment if that helps. We didn't used HAL atm. I can say much more if we begin to use HAL in our services. I think all solutions to map a pattern to an url can help us. I can not say much more at the moment. I hope we use HAL soon and we can try to implement such a behaviour soon.

**Comment by andreasevers** [comment]
We have used HAL all over the place in our microservices architecture, and we just started adopting pact.
We have exposed an index resource on every microservice which exposes a bunch of links. This is easily done with spring-hateoas.

Now I am thinking of just putting the rel as "path" in the pact json, and then in the producer use the reflection abstraction of spring-hateoas to look up the link and its corresponding url in the index resource (or something along those lines).

Path variables and request parameters can be put in the "query" map and can be dynamically substituted in the retrieved url.

I will see how we can extend this library to use the index exposed in a microservice producer.

**Comment by uglyog** [comment]
The generators that are part of (V3)[https://github.com/pact-foundation/pact-specification/tree/version-3#introduce-example-generators] could be used to deal with URLs. The generator could provide the correct URLs for where the test is running, and then a matcher could deal with the provider returning URLs specific to itself.

**Comment by uglyog** [comment]
I've finally come back to this issue, and have done some testing using the pact broker HAL and the Pact-JVM Pact Broker client (which navigates in a similar way to Traversion). Relative paths work without any issue, but if absolute paths are returned (as with the pact broker), this is an issue.

I have added URL support to the consumer DSLs that take a base path and a list of path fragments. The full URL is used for the example bodies and a regular expression matcher is used for verifying the provider URLs. I have tested this successfully against the actual pact broker (which means I verified a Pact against the Pact broker itself :-D )

See the example test here: https://github.com/DiUS/pact-jvm/blob/master/pact-jvm-provider/src/test/groovy/au/com/dius/pact/provider/broker/PactBrokerClientSpec.groovy#L147
